File: frontend/src/api/timeline.ts
import { api } from './client'
import type { EntryKind, PaidBy, PayMethod } from './entries'
import type { PayPurpose } from './payments'

export interface TimelineEntry {
  type: 'entry'
  id: number
  kind: EntryKind
  name: string
  quantity: string
  unit: string
  unit_label: string | null
  unit_price: string
  amount: string
  paid_by: PaidBy
  payment_method: PayMethod | null
  is_rework: boolean
  has_receipt: boolean
  note: string | null
  entry_date: string
  created_at: string
}

export interface TimelinePayment {
  type: 'payment'
  id: number
  kind: 'payment'
  purpose: PayPurpose
  amount: string
  method: PayMethod
  note: string | null
  paid_at: string
  created_at: string
}

export type TimelineItem = TimelineEntry | TimelinePayment

export interface TimelinePage {
  items: TimelineItem[]
  page: number
  pages: number
  total: number
}

export type TimelineKind = EntryKind | 'payment'

export function listTimeline(params: {
  projectId: number
  kind?: TimelineKind
  dateFrom?: string
  dateTo?: string
  page?: number
}) {
  const qs = new URLSearchParams({ page: String(params.page ?? 1) })
  if (params.kind) qs.set('kind', params.kind)
  if (params.dateFrom) qs.set('date_from', params.dateFrom)
  if (params.dateTo) qs.set('date_to', params.dateTo)
  return api<TimelinePage>(
    `/api/projects/${params.projectId}/timeline?${qs}`,
  )
}

File: backend/app/schemas/timeline.py
"""Tarix lentasi — yozuvlar va to'lovlar bitta oqimda."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel


class TimelineEntry(BaseModel):
    type: Literal["entry"] = "entry"
    id: int
    kind: str  # work | material | expense
    name: str
    quantity: Decimal
    unit: str
    unit_label: str | None
    unit_price: Decimal
    amount: Decimal
    paid_by: str
    payment_method: str | None
    is_rework: bool
    has_receipt: bool
    note: str | None
    entry_date: date
    created_at: datetime


class TimelinePayment(BaseModel):
    type: Literal["payment"] = "payment"
    id: int
    kind: Literal["payment"] = "payment"  # filtr birxilligi uchun
    purpose: str  # labor | material
    amount: Decimal
    method: str
    note: str | None
    paid_at: date
    created_at: datetime


TimelineItem = TimelineEntry | TimelinePayment


class TimelinePage(BaseModel):
    items: list[TimelineItem]
    page: int
    pages: int
    total: int
